proposals/gc/Overview.md

```diff
@@ -1 +1,728 @@
+  (local $env (ref $inner-env))
+  (block $fail (result (ref anyref))
+    (set_local $env (cast_down anyref (ref $inner-env) $fail (get_local _$anyenv)))
+    (get_local $y)
+    (get_field $inner-env $a (get_local $env))
+    (f64.add)
+    (get_field $inner-env $x (get_local $env))
+    (f64.add)
+    (return)
+  )
+  (unreachable)
+)
+
+(func $caller (result f64)
+  (local $clos (ref $clos-f64-f64))
+  (set_local $clos (call $outer (f64.const 1)))
+  (call_ref
+    (get_field $clos-f64-f64 $env (get_local $clos))
+    (f64.const 2)
+    (get_field $clos-f64-f64 $code (get_local $clos))
+  )
+)
+```
+
+Needs:
+* function pointers
+* universal type of references
+* down casts
+
+The down cast for the closure environment is necessary because expressing its static type faithfully would require first-class generics and existential types.
+
+An alternative is to provide [primitive support](#closures) for closures.
+
+
+### Parametric Polymorphism
+
+TODO: via type `anyref` and `intref`
+
+
+### Type Export/Import
+
+* Want to allow type definitions to be imported from other modules.
+* As much as possible of the above constructions should be allowed with abstract types.
+* More complicated linking patterns might require user-space linking hooks.
+* Possibly: allow abstract type exports (encapsulation)?
+* Lots of tricky details here, mostly ignore for now...
+
+
+## Basic Functionality: Simple Aggregates
+
+* Extend the Wasm type section with new constructors for aggregate types.
+* Extend the value types with new constructors for references and interior references.
+* Aggregate types are not value types, only references to them are.
+* References are never null; nullable reference types are separate.
+
+
+### Structures
+
+*Structure* types define aggregates with heterogeneous fields that are _statically indexed_:
+```
+(type $time (struct (field i32) (field f64)))
+(type $point (struct (field $x f64) (field $y f64) (field $z f64)))
+```
+Such types can be used by forming *reference types*, which are a new type of value type:
+```
+(local $var (ref $point))
+```
+
+Fields are *accessed* with generic load/store instructions that take a reference to a structure:
+```
+(func $f (param $p (ref $point))
+  (store_field $point $y (get_local $p)
+    (load_field $point $x (get_local $p))
+  )
+)
+```
+All accesses are type-checked at validation time.
+
+Structures are *allocated* witha  `new` instruction that take initialization values for each field.
+The operator yields a reference to the respective type:
+```
+(func $g
+  (call $g (new_struct $point (i32.const 1) (i32.const 2) (i32.const 3)))
+)
+```
+Structures are garbage-collected.
+
+Structures can be compared for identity:
+```
+(same (new $point ...) (new $point ...))  ;; false
+```
+TODO: Could even allow heterogeneous equality (equality between operands of different type), but that might lead to some discontinueties or even prevent some potential optimizations?
+
```

Comment:
Right. Not currently included, but may be a worthwhile extension if there turn out to be sufficiently many use cases of large structures that need to be copied by value.